Source/DafnyStandardLibraries/src/Std/Actions/Enumerators.dfy

```diff
@@ -0,0 +1,292 @@
+      && Outputs(history) == Seq.Map(x => Some(x), elements[..values]) + Seq.Repeat(None, nones)
+    }
+
+    method Invoke(t: ()) returns (value: Option<T>)
+      requires Requires(t)
+      reads Reads(t)
+      modifies Modifies(t)
+      decreases Decreases(t).Ordinal()
+      ensures Ensures(t, value)
+    {
+      assert Requires(t);
+
+      if |elements| <= index {
+        value := None;
+      } else {
+        value := Some(elements[index]);
+        index := index + 1;
+      }
+      Update((), value);
+      // TODO: Doable but annoying
+      assume {:axiom} CanProduce(history);
+      assert Valid();
+    }
+
+
+    method RepeatUntil(t: (), stop: Option<T> -> bool, ghost eventuallyStopsProof: ProducesTerminatedProof<(), Option<T>>)
+      requires Valid()
+      requires eventuallyStopsProof.Action() == this
+      requires eventuallyStopsProof.FixedInput() == t
+      requires eventuallyStopsProof.StopFn() == stop
+      requires forall i <- Consumed() :: i == t
+      reads Repr
+      modifies Repr
+      decreases Repr
+      ensures Valid()
+    {
+      DefaultRepeatUntil(this, t, stop, eventuallyStopsProof);
+    }
+
+    ghost function Limit(): nat {
+      |elements|
+    }
+
+    lemma ProducesTerminated(history: seq<((), Option<T>)>)
+      requires Action().CanProduce(history)
+      requires (forall i <- Inputs(history) :: i == FixedInput())
+      ensures exists n: nat | n <= Limit() :: Terminated(Outputs(history), StopFn(), n)
+    {
+      assert Terminated(Outputs(history), StopFn(), |elements|);
+    }
+
+  }
+
+  trait Pipeline<U, T> extends Enumerator<T> {
+
+    const upstream: Enumerator<U>
+    const buffer: Collector<T>
+
+    ghost predicate Valid()
+      reads this, Repr
+      ensures Valid() ==> this in Repr
+      ensures Valid() ==> CanProduce(history)
+      decreases height, 0
+    {
+      && this in Repr
+      && ValidComponent(upstream)
+      && ValidComponent(buffer)
+      && upstream.Repr !! buffer.Repr
+      && CanProduce(history)
+    }
+
+    // TODO: needs refinement
+    ghost predicate CanProduce(history: seq<((), Option<T>)>)
+      decreases height
+    {
+      true
+    }
+
+    method Invoke(t: ()) returns (r: Option<T>)
+      requires Requires(t)
+      reads Repr
+      modifies Modifies(t)
+      decreases Decreases(t).Ordinal()
+      ensures Ensures(t, r)
+    {
+      label start:
+      while (|buffer.values| == 0)
+        invariant upstream.Repr !! buffer.Repr
+        invariant fresh(Repr - old(Repr))
+        invariant fresh(buffer.Repr - old(buffer.Repr))
+        invariant fresh(upstream.Repr - old(upstream.Repr))
+        invariant Valid()
+        invariant buffer.ValidAndDisjoint()
+        invariant upstream.ValidAndDisjoint()
+        invariant history == old(history)
+        modifies Repr
+        decreases upstream.Remaining()
+      {
+        var u := upstream.Next();
+        Repr := {this} + upstream.Repr + buffer.Repr;
```

Comment:
Alternatively, it may work to do `Repr := Repr + upstream.Repr;`. Or, if this kind of assignment was done often, you could put the assignment, as you have it, into a helper ghost method.